### ISSUE TITLE
Remove TDZ scopes

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#10245]
> I'm sorry that I don't follow issue templates.

From #10227.

```js
for (var key in obj)
    ;
```

In this case, `eslint-scope` makes a TDZ scope around `obj`. The TDZ scope steals reference objects in `ForInStatement#right` then it has caused bugs. TDZ is the word about a timing matter rather than a structure matter, so it's wrong that `eslint-scope` makes TDZ scopes. Actually, all our core rules just ignore the TDZ scopes.

We should remove the TDZ scopes.

However, I think that we should take the removal of TDZ scopes as a breaking change because the TDZ scope has existed in our variable analysis while over 3 years ago.

**Comment by Aladdin-ADD** [comment]
It would be a core change,  so will it be discussed in the next TSC meeting?
Temporaryly labeled `tsc`, but feel free to remove if no need. 😏

**Comment by mysticatea** [comment]
Yes, it's necessary.

**Comment by ilyavolodin** [comment]
@mysticatea Could you explain in more details would would happen with this following code if TDZ is removed?
```js
for (let key in obj) {
  const a = obj[key];
}
```
As far as I know, content of the `for` loop should be inside TDZ, removing it will produce incorrect results in scope analysis. Or are you just saying that we have a bug where loop statement is also treated as part of TDZ that needs fixing?

**Comment by mysticatea** [comment]
> content of the for loop should be inside TDZ

I'm not sure about it. Temporal Dead Zone is a timing matter (`Temporal` means a thing about time). I.e., if it touches a certain variable before that variable declaration executes then it throws a ReferenceError. For example:

```js
function foo() {
    a
}

foo() // throw a ReferenceError because `let a` hasn't executed. This is in TDZ.
let a
foo() // good work because `let a` has executed. This is not in TDZ.
```

In the following code:

```js
for (let key in obj) {
  const a = obj[key];
}
```

TDZ doesn't exist because we cannot touch the `key` variable before the variable declaration executes.

**Comment by ilyavolodin** [comment]
Ok, I think the confusion that I have is in terminology. You are completely correct that TDZ only applies to execution, because it's based on timing (or rather, not timing, but currently executing line of code). So TDZ is a wrong term to use. However, the following code is invalid, because of the block scoping issues:

```js
for (let key in obj) {
  const a = obj[key];
}
console.log(a);
```
Would removal of TDZ that you are proposing also remove block scoping? Because if it does, I think that's incorrect.

**Comment by mysticatea** [comment]
> Would removal of TDZ that you are proposing also remove block scoping?

Of course, no. I'm talking about [the TDZScope class in eslint-scope package](https://github.com/eslint/eslint-scope/blob/982a71f9430dc9c5de78a3923f4d6dc2243a3386/lib/scope.js#L634-L638). The removal of that doesn't change scope structure. We are just ignoring that scope.

In the following code:

```js
for (let key in obj) {
  const a = obj[key];
}
console.log(a);
```

Currently `eslint-scope` makes the following structure:

```json
{
    "type": "global",
    "variables": [],
    "references": [ "console", "a" ],
    "childScopes": [
        {
            "type": "for",
            "variables": [ "key" ],
            "references": [],
            "childScopes": [
                {
                    "type": "TDZ",
                    "variables": [],
                    "references": [ "obj" ],
                    "childScopes": []
                },
                {
                    "type": "block",
                    "variables": [ "a" ],
                    "references": [ "obj", "key" ],
                    "childScopes": []
                }
            ]
        }
    ]
}
```

After it removes TDZ scopes:

```json
{
    "type": "global",
    "variables": [],
    "references": [ "console", "a" ],
    "childScopes": [
        {
            "type": "for",
            "variables": [ "key" ],
            "references": [ "obj" ],
            "childScopes": [
                {
                    "type": "block",
                    "variables": [ "a" ],
                    "references": [ "obj", "key" ],
                    "childScopes": []
                }
            ]
        }
    ]
}
```

**Comment by ilyavolodin** [comment]
Ah, I see. Thanks for the explanation. In that case, I'm in support of this change (I think). I do think it has potential to break some of the rules, though. So we need to be careful about introducing it.

**Comment by mysticatea** [comment]
**TSC Summary**: ESLint should not make TDZ scopes because TDZ is a timing matter. TDZ scopes have caused a bug (#10227). The removal does not have any effect to core rules, at least in our tests (#10270).
**TSC Question**: Should we accept this change -- remove TDZ scopes?

**Comment by not-an-aardvark** [comment]
This issue was accepted in [today's TSC meeting](https://github.com/eslint/tsc-meetings/blob/c271eee0c355058b90acb511a6e70a0c4520b121/notes/2018/2018-04-26.md#remove-tdz-scopes).
